src/unicode/capi.h

```diff
@@ -0,0 +1,76 @@
+/**
+ * This file is part of the "libunicode" project
+ *   Copyright (c) 2021 Christian Parpart <christian@parpart.family>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <stddef.h>
+#include <stdint.h>
+
+#if !defined(__cplusplus)
+extern "C" {
+#endif
+
+/// UTF-8 character or 8bit segment of an UTF-8 character.
+typedef char u8_char_t;
+
+/// UTF-32 codepoint.
+typedef uint32_t u32_char_t;
+
+/**
+ * Counts the number of grapheme clusters for given sequence of codepoints.
+ *
+ * Use this function to determine the number of
+ * user perceived characters (grapheme clusters).
+ *
+ * @param codepoints   pointer to the first codepoint.
+ * @param n            number of codepoints to count the grapheme clusters for.
+ *
+ * @return number of user perceived characters (grapheme clusters) counted
+ *         in [codepoints, codepoints+n).
+ */
+int u32_gc_count(u32_char_t const* codepoints, size_t n);
+int u8_gc_count(u8_char_t const* codepoints, size_t n);
```

Comment:
Oh wow, thats darn simple and will nicely work for finite string input (where a string is known to be finished) :+1: 

But what about a stream interface? I remember having that issue while playing around with the grapheme PR, and kinda had to daisy-chain the last grapheme classification and its offset with the first codepoint of the next chunk.

Reply:
for streaming interfaces you can do it yourself using `bool u32_grapheme_unbreakable(a, b)`.

basically like that:
```cpp
int count = 0;
u32_char_t carry = 0;
// ...
void streaming_input(u32_char_t _nextCodepoint)
{
    if (!u32_grapheme_unbreakable(carry, _nextCodepoint))
        count++;            // you have found a boundary
    carry = _nextCodePoint; // remember last codepoint being processed
    // ...

    // count here always contains the number of already and *fully* processed grapheme clusters.
}
```

Creating a helper function and would be trivial. I am not yet sure that is needed though :)

Reply:
Yeah true, in the ends it is the question how high level you want to go with the lib. Lets peeps implement it themselves it fine imho (since it is that easy).

Reply:
As high-level as needed to convince the terminal world for good. :-)